### PR TITLE
New library format

### PR DESCRIPTION
**Pull request by axlt2002**
By changing the Alfa library format from .nfo to .info, the interworking with the Kodi library is now possible (i.e. a TV Show can now have both local and streaming episodes).

### CODE REVIEW

File: plugin.video.alfa/channels/videolibrary.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 
 import os
+import xbmc
```
Comment (alaquepasa) [inline plugin.video.alfa/channels/videolibrary.py]: ¿Esto no romperá la versión mediaserver?

File: plugin.video.alfa/channels/videolibrary.py [hunk 2]
```diff
@@ -10,10 +11,43 @@
...
+            counter += 1
+            p_dialog.update((counter*100)/num_nfos)
+            os.rename(filename, filename.replace('.nfo', '.info'))
+
+        for filename in tvshows_nfos:
+            counter += 1
+            p_dialog.update((counter*100)/num_nfos)
+            os.rename(filename, filename.replace('.nfo', '.info'))
+
+        xbmc.sleep(2000)
```
Comment (alaquepasa) [inline plugin.video.alfa/channels/videolibrary.py]: Si peta la mediaserver tal vez lo podrías poner como platformtools y en xbmc se pone está línea y en mediaserver pones un sleep

File: plugin.video.alfa/channels/videolibrary.py (hunk 1 above)
Comment (axlt2002) [reply plugin.video.alfa/channels/videolibrary.py]: I do not understand the point. I think that "import xbmc" is arelady used in other modules of Alfa, isn't it?

File: plugin.video.alfa/channels/videolibrary.py (hunk 2 above)
Comment (axlt2002) [reply plugin.video.alfa/channels/videolibrary.py]: See my previous comment. Btw, that sleep is used just to keep the notification on the screen for two seconds after the library conversion is finished. It is not strictly necessary then.

File: plugin.video.alfa/channels/videolibrary.py (hunk 1 above)
Comment (alaquepasa) [reply plugin.video.alfa/channels/videolibrary.py]: Not so much, and all they are broken:
```Traceback (most recent call last):
  File "/home/osmc/bin/alfa/WebSocket.py", line 68, in run
    controller.run(path)
  File "/home/osmc/bin/alfa/platformcode/controllers/html.py", line 50, in run
    launcher.run(item)
  File "/home/osmc/bin/alfa/platformcode/launcher.py", line 108, in run
    exec "itemlist = channelmodule." + item.action + "(item)"
  File "<string>", line 1, in <module>
  File "/home/osmc/bin/alfa/channels/tvsinpagar.py", line 892, in findvideos
    import xbmc
ImportError: No module named xbmc
```
If you don't want to break non Kodi versions (now only Mediaserver, in the past you have Plex too) you can't use import xbmc in common features, like a channel.

**Comment by axlt2002** [comment]
I have reported the reasons for the modifications in the pull request. The documentation you are referring to says also that .nfo files have a specific format...and Alfa uses a format different from the one of Kodi; so is not just a question of file extension. There are then two solutions in order to avoid conflicts:

1. Change the format of the Alfa .nfo file to the format used by Kodi -> very long and complicated, with malfunctioning issues (is not possible to store the information of the channel)

2. Just change the extension to .info -> very quick and simple, without risks (the format of the file remains the same)

Having two different extensions (.nfo for Kodi and .info for Alfa) enables the possiblity to have mixed content in a TV Show, where you can have local stored episodes as well as stream episodes through the Kodi library. Acutally, if you use the Kodi library and decide to export it, all the .nfo files of Alfa are overwritten and than they are no more usable by Alfa, providing an error when trying to play an episode. If Alfa uses its hown extension .info, the issue above doesn't happen.

**Comment by alaquepasa** [comment]
But what I know is that alpha use the feature:

> When Kodi locates the nfo file it will load the *.nfo file and:
>     - If the file is a parsing nfo, Kodi will parse the nfo file for a direct URL to the scraper info page, then scrape the online metadata. See: Parsing NFO

This is why the first line of the file is an tmdb URL (the rest of the is for Alfa addon, not for Kodi). This still works if you rename it to .info?

In other works, is like have:
tvshow.nfo with online URL
alfa_metadata.json with metadata for Alfa

But instead of have 2 files, they can add the alfa_metadata.json in tvshow.nfo because Kodi will still work with the first URL.

File: plugin.video.alfa/channels/videolibrary.py (hunk 1 above)
Comment (axlt2002) [reply plugin.video.alfa/channels/videolibrary.py]: May be I'm missing a point but making a search inside the files in plugin.video.alfa, there are a lot of results of the string "import xbmc", for example, downloader.py, videolibrarytools.py, etc. Are these modules not used by Mediaserver version?

Comment (alaquepasa) [reply plugin.video.alfa/channels/videolibrary.py]: In your "downloader" example you can see that they are in a try..catch and in "videolibrarytools" it is inside of:
    if config.is_xbmc():

Comment (alaquepasa) [reply plugin.video.alfa/channels/videolibrary.py]: In your code is very easy to do platform independent. You only want "import xbmc" to performa  sleep.
You can do some versions:
Implement a platform code "sleep" where XMBC does a xbmc.sleep and others a time.sleep()

or simply put in your code:
try:
    import xbmc
    xbmc.sleep
catch:
    time.sleep

or without exceptions
if config.is_xbmc():
import xbmc
xbmc.sleep()
else:
time.sleep

Comment (axlt2002) [reply plugin.video.alfa/channels/videolibrary.py]: Thanks for further interact on this point. I'm starting to understand but what it is still unclear to me is that the is_xbmc() function inside config.py returns always True...

**Comment by axlt2002** [comment]
Thanks a lot for your explaination alaquepasa. I was not aware of the Parsing NFO (such a smart solution!).

Making one step back, my intention was to make Alfa able to manage mixed content (local and streaming episodes for TV Shows). What brought me to go for the proposed solution of chaning the extension was the impossiblity to choose which scraper use for TV Shows in the automatic Kodi library configuration provided by Alfa. Actually, if I'm not wrong, Alfa checks if the TMDB scraper for TV Shows is installed; if the answer is not, the scraper is downloaded (and setting.xml with default parameters is created). Despite this, when Alfa adds the TV Show folder as Kodi source, the TVDB is the scraper that is always set (and without the possilbity to change the parameters). I'm actually consider to modify this part of the code inside xbmc_videolibrary.py for a more flexible solution: choose the scraper between TVDB and TMDB and the possilbity to modify the parameters. What do you think?

Having said all this, I have the two following considerations:

1. If the users decides to export the Kodi library, Alfa .nfo files can be overwritten. This will corrupt Alfa behaviour because now the .nfo files are in the Kodi format. May be this use case is quite rare...but at the same time I don't know how such issue can be managed. Ideas?
2. I have noticed that when Alfa is uninistalled, Kodi maintains the videolibrary and download folders in its sources. Do yo think that there is a solution to remove them when Alfa is uninstalled?

Thanks a lot again for your support!

P.S.: Which is the background for naming the folders as "Title [imdb id]" instead of the Kodi format "Title (year)"?

File: plugin.video.alfa/channels/videolibrary.py (hunk 1 above)
Comment (alaquepasa) [reply plugin.video.alfa/channels/videolibrary.py]: Yes, here is always true: https://github.com/alfa-addon/addon/blob/master/plugin.video.alfa/platformcode/config.py#L57

But here is always false: https://github.com/alfa-addon/addon/blob/master/mediaserver/platformcode/config.py#L32

It is normal that "is_xbmc" is always true in XMBC. And false in others (like mediaserver).

Comment (axlt2002) [reply plugin.video.alfa/channels/videolibrary.py]: Thanks for the reply alaquepasa. my initial understanding was that Alfa could be installed over Kodi and Mediserver...now I understood that there are actually two different add-on that maybe share a lot of code as well as the name of the modules but with customized functions. All clear now; I will implement the code you suggested.

**Comment by axlt2002** [comment]
I'm going to review the modifications according to the discussion above.

Thanks for the collaboration.

**Comment by alaquepasa** [comment]
I know that autoinstall always configure TVDB.

The addon have hardcoded TVDB settings, but not TMDB.

I tryed to modify this, but I already have my tv shows folder with TMDB (going to Kodi -> Videos and changing the folder content scraper manually) i didn't continue looking this :laughing:  When I look this what I did was a dict with the settings:
scraperSettings = { 'plugin.scraper.tvdb' = "settings", "plugin.scraper.tmdb" = "settings" }
And in other place have an string with the current scraper that will use:
scraper = "plugin.scraper.tvdb" if option == 1 else "plugin.scraper.tmdb"

For example, so then you can insert "scraper" and scraperSettings[scraper]

For your first point: I don't know, i never exported my Kodi library. I only use Kodi with Alfa and I never exported this (when I want to move the library to other site I copy my Alfa -> videolibrary folder maintaining only the tvshow.nfo files, Alfa will recreate all with this).

For 2, I don't know if Kodi have some "uninstall" callback, I supose that he will not have.

Fir PS: I don't know.

**Pull request closed without merging**